Spatial Rule Lookup (e.g. Country Specific Rules)

boldtrn opened the pull request:
This PR intents to solve #175.

The basic data structure is an array, that separates the world (or an area) into tiles, with a given resolution. Each tile contains a reference to a set of rules.

Rules contain a set of polygons in which their rules are valid. Using an array has to disadvantages: (1) covering unused space, e.g. water and (2) inexact resoultion. Currently the default resolution about 20*20km at the equator (0.1 degree). We can achieve exact results when doing a Polygon contains check, which can be configured.

Right now I prepared two set of rules, Austria and Germany, but rules can be written for every Country. All countries are available in the countries.geo.json file. This file is based on the admin levels from [NaturalEarthData](http://www.naturalearthdata.com/) under [public domain](http://www.naturalearthdata.com/about/terms-of-use/). As the file would be 25mb, is reduced it to 10% using DP. Therefore the borders are sometimes slightly inaccurate (should not be too bad, otherwise let me know).

Lookup speed is not too bad. 10 000 000 lookups in an area that is filled approximately half by a polyong, takes between 3-4 seconds on my machine.

TODOs:

- [x] Nr of rules is currently limited to 255, but since we have to create a new rule for every unique set of country borders, the array might be with 20-30 countries, depending on the borders.
- [x] The SpatialLookup is currently created in the web module. It would be better if we could have it as `Storable<LocationIndex>`. That way we could load it, once it was created (even though creation is quite fast). Also I am not sure how to do this (especially with the rule reference array, the lookup array should be easy)?
- [x] How to evaluate these rules? Currently I do this in the DataFlagEncoder, however it might be better to do this in the GenericWeighting? IMHO the DataFlagEncoder is representing the tags of a given source (e.g. OSM), but not business logic. For example we are also not saving the default max speeds in the edge, but calculate it every time in the GenericWeighting
- [ ] Consider using the exact country borders - would slow down the Polygon contains approx. 10 times. Also we have to host this file, as IMHO 25mb should not be put into the repo.
- [x] Differentiate between Polygon and Multipolygon? -> Later
- [ ] Add Javadoc to the Interfaces
- [x] Make spatial index configureable. What rules to load and if it should be used at all.

Let me know what you think. Currently this is still work in progress, so any comments are more than welcome.

karussell commented:
Thanks - will have a look!

> As the file would be 25mb, is reduced it to 10% using DP

Maybe this is not too critical. How large is the 25mb file if compressed? As git and jar normally does compression and it will be big just on the disc and not in production.

boldtrn commented:
> How large is the 25mb file if compressed?

2.5mb. You can view the file here: http://bl.ocks.org/d/7280192619e0cc25f609dafe5c0de2cf

I think the borders are not too bad, but we can definetly use the exact borders if this becomes an issue.

karussell commented on core/src/main/java/com/graphhopper/GraphHopper.java:
```diff
@@ -1261,4 +1265,13 @@ public void setNonChMaxWaypointDistance(int nonChMaxWaypointDistance) {
         this.nonChMaxWaypointDistance = nonChMaxWaypointDistance;
     }
 
+    public void setSpatialRuleLookup(SpatialRuleLookup spatialRuleLookup) {
```
Hmmh, I do not like this procedure much. We should not provide a setter here, if this is not an additional thing for GraphHopper (infact GraphHopper.spatialRuleLookup is unused).

Instead the user should/could do `graphHopper.getEncodingManager().getEncoder("generic").setXY` ... or maybe independent on generic and just `graphHopper.getEncodingManager().setXY`?

boldtrn replied:
Makes sense :+1: I think it makes sense to start only with generic for now. We can always extend that.

karussell commented on core/src/main/java/com/graphhopper/routing/util/DataFlagEncoder.java:
```diff
@@ -217,6 +224,15 @@ int getAccessValue(ReaderWay way) {
+                accessValue = getEdgeValueForAccess(rule.isAccessible(way, ""));
+            }
+
+        }
```
Can you explain this case?

boldtrn replied:
Yes, the idea was to first check the ReaderWay if there is an access restriction value on it. If not check the country rules to eventually add a restriction.

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/Polygon.java:
```diff
@@ -0,0 +1,173 @@
+ *
+ * @author Robin Boldt
+ */
+public class Polygon {
```
This needs more javadocs

boldtrn replied:
Added some, probably still needs more :)

karussell commented on core/src/main/java/com/graphhopper/routing/util/DefaultFlagEncoderFactory.java:
```diff
@@ -17,7 +17,15 @@
+import com.graphhopper.util.shapes.BBox;
+
+import java.util.ArrayList;
+import java.util.List;
```
Not necessary (?)

boldtrn replied:
Yes, removed.

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRule.java:
```diff
@@ -0,0 +1,42 @@
+ */
+public interface SpatialRule {
+
+    int getMaxSpeed(ReaderWay readerWay, String transportationMode);
```
return type should be double

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRule.java:
```diff
@@ -0,0 +1,42 @@
+ *
+ * @author Robin Boldt
+ */
+public interface SpatialRule {
```
Hmmh, I'm not sure about the interface. It looks like we'll break implementations as soon as we need a different data from the index. Instead of getCountryIsoA3Name, isAccessible and getMaxSpeed maybe we should be more generic here? (`Double getData("max speed", Double.class)`) hmmh, we'll have to think about this.

boldtrn replied:
Yes that is true. That might be better. Would make it way more flexible to change it.

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRuleContainer.java:
```diff
@@ -0,0 +1,78 @@
+    @Override
+    public boolean equals(Object o){
+        if(o instanceof SpatialRuleContainer){
+            if(this.represents(((SpatialRuleContainer) o).getRules()))
```
This is not really correct. Only because X has all rules of Y does not mean the opposite i.e. they are not necessarily equal. But maybe this is not important?

boldtrn replied:
If X and Y have the same number of rules, than this should be correct :)

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRuleContainer.java:
```diff
@@ -0,0 +1,78 @@
+        return this.rules.iterator().next();
+    }
+
+    public boolean represents(Set<SpatialRule> rules) {
```
naming it "contains" would make it more explicit (?)

boldtrn replied:
I thought about that, This is a rulesEquals and not a contains.

boldtrn replied:
There was a reason that I extracted that method from the equals method. But I cannot remember it right now.

karussell replied:
Yes, you are right, but it is slightly confusing. Why not call the equals of the underlying collection?

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRuleLookup.java:
```diff
@@ -0,0 +1,39 @@
+
+    void addRule(SpatialRule rule);
+
+    SpatialRule getEmptyRule();
```
I would remove this method and make the EMPTY_RULE constant part of this interface

boldtrn replied:
Makes sense. Changed that.

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRuleLookup.java:
```diff
@@ -0,0 +1,39 @@
+ *
+ * @author Robin Boldt
+ */
+public interface SpatialRuleLookup {
```
javadocs are missing

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/countries/DefaultSpatialRule.java:
```diff
@@ -0,0 +1,78 @@
+        String highwayTag = readerWay.getTag("highway", "");
+
+        // As defined in: https://wiki.openstreetmap.org/wiki/OSM_tags_for_routing/Access-Restriction
+        // We tried to find generally forbidden tags
```
forbidden for a certain transportationMode?

boldtrn replied:
Yes, currently the SpatialRule only uses car rules, but should be used for other transportation modes as well.

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/countries/DefaultSpatialRule.java:
```diff
@@ -0,0 +1,78 @@
+/**
+ * Default Implementation for the SpatialRule that contains the current default Values
+ */
+public class DefaultSpatialRule extends AbstractSpatialRule {
```
Shouldn't this be part of the AbstractSpatialRule, e.g. getCountryIsoA3Name implementation is still missing and so this is not really a default (?)

boldtrn replied:
Not sure. The idea was to define default rules for streets that are found frequently in different countries. Adding a default CountryIsoA3Name might now work? I might make sense to make this class abstract as well. The idea was to provide defaults, that why I called it Default. Should we move it to the AbstractSpatialRule?

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/countries/GermanySpatialRule.java:
```diff
@@ -0,0 +1,60 @@
+            case "motorway":
+                return Integer.MAX_VALUE;
+            case "trunk":
+                return Integer.MAX_VALUE;
```
Let us define something for this. E.g. there are 130 limits but I think somewhere I used 140 ... the problem with MAX_VALUE is that this cannot be stored in the encoder

boldtrn replied:
The idea was to use this as an upper bound. If the Encoder tries to initialize the way with a bigger value we are overwriting it with the Spatial MaxSpeed. This leaves most of the responsibility with the Encoder. Should I change it to 140?

karussell replied:
This is fine but should be probably explicitly mentioned in the javadocs of the interface?

karussell commented on core/src/test/java/com/graphhopper/routing/util/spatialrule/SpatialRuleLookupArrayTest.java:
```diff
@@ -0,0 +1,203 @@
+        assertEquals(AccessValue.EVENTUALLY_ACCESSIBLE, spatialRuleLookup.lookupRule(48.210033, 16.363449).isAccessible(livingStreet, ""));
+    }
+
+    private Polygon parsePolygonString(String polygonString){
```
You could use JSONObject in tests to deserialize the string.

boldtrn replied:
Do you think this would be the better approach? I tried to keep the core clean of JSON.

karussell commented on core/src/main/java/com/graphhopper/json/geo/GeoJsonPolygon.java:
```diff
@@ -0,0 +1,77 @@
+/**
+ * Created by robin on 12.12.16.
+ */
+public class GeoJsonPolygon implements Geometry {
```
Javadocs are missing.

karussell commented on core/src/main/java/com/graphhopper/json/geo/GeoJsonPolygon.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Override
+    public String getType() {
+        return null;
```
Here we should return Polygon or MultiPolygon or create two implementations?

boldtrn replied:
Added Polygon for now. We definitely have to think about how we handle this in the future.

karussell commented on web/src/main/resources/com/graphhopper/http/README.MD:
```diff
@@ -0,0 +1,5 @@
+
+The Geojson has roughly 25MB, so I reduced it with http://www.mapshaper.org/ using Douglas Peuckar at 10%. The resulting file is 2.4mb.
+
+The rules are taken from the OSM Wiki: ttps://wiki.openstreetmap.org/wiki/OSM_tags_for_routing/Access-Restrictions
```
Should be in the NOTICE.md instead (?)

And why is this and countries.geo.json in package com/graphhopper/http ?

boldtrn replied:
> Should be in the NOTICE.md instead (?)

Yes thats correct. I wanted to fix this once we have a final location for the countries.geo.json.

> And why is this and countries.geo.json in package com/graphhopper/http ?

Currently I load the countries.geo.json with `SpatialRuleLookupBuilder.class.getResource("countries.geo.json")` and SpatialRuleLookupBuilder is in com.graphhopper.http. I had it in the core first, but wasn't able to read the file in all cases (I think when running the tests with Maven).

karussell commented on web/src/main/java/com/graphhopper/http/DefaultModule.java:
```diff
@@ -52,6 +52,9 @@ public GraphHopper getGraphHopper() {
         GraphHopper tmp = new GraphHopperOSM().forServer().init(args);
+        //TODO We should set the BBox and resolution
+        //TODO, move to a more appropriate place
+        tmp.setSpatialRuleLookup(SpatialRuleLookupBuilder.build());
```
related to the TODOs :) ... it is suboptimal that it cannot be configured where to load the json from and the resolution ...

boldtrn replied:
Yes, I know :) I should fix this.

karussell commented on web/src/main/java/com/graphhopper/http/SpatialRuleLookupBuilder.java:
```diff
@@ -0,0 +1,84 @@
+            GHJson ghJson = new GHJsonBuilder().create();
+            JsonFeatureCollection jsonFeatureCollection = ghJson.fromJson(new FileReader(new File(SpatialRuleLookupBuilder.class.getResource("countries.geo.json").getFile())), JsonFeatureCollection.class);
+
+            // TODO find outer Border of all used features and create SpatialRuleLookupArray onlyo for these Bounds
```
What do you mean here?

(+minor spelling issues)

boldtrn replied:
The idea was to calculate a BBox for all SpatialRules. Initializing an Array for world wide coverage makes no sense If we only have a rule for a subset of the world (e.g., Germany and Austria). Would be quite easy to do and would save quite some storage.

karussell commented on web/src/main/java/com/graphhopper/http/SpatialRuleLookupBuilder.java:
```diff
@@ -0,0 +1,84 @@
+                }
+            }
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
```
please never catch exceptions and continue silently. Instead rethrow e.g. via `new RuntimeException(e)`

boldtrn commented on core/src/main/java/com/graphhopper/json/geo/Geometry.java:
```diff
@@ -17,11 +17,12 @@
 
 /**
- * This interface is used to define an area or location being used to block or change areas in the graph.
+ * This interface is used to define an area or location in a geojson file.
```
Oups, you'r right :)

karussell commented:
I tested this with the landmarks branch and this seems to work (at least the spatial rule part, still something else still does not work ;))

Some minor changes I had to do or would like to have :) (maybe already done, did not review your latest changes too detailed)

 * calculate boundaries from the polygons instead of providing one
 * https://github.com/graphhopper/graphhopper/commit/2c04f4790eb646246fa47fced2fd4f7c24f054cc
 * https://github.com/graphhopper/graphhopper/commit/8dc447d5785953e4513562c954c46ec693663289

boldtrn commented:
> I tested this with the landmarks branch

Thanks for testing :)

> seems to work (at least the spatial rule part, still something else still does not work there ;))

But unrelated to this PR?

> Some minor changes I had to do or would like to have :)

Thanks for the two commits. I will add these to this branch as well. Calculating the boundaries from the polygons is already on my TODO list as well.

> (maybe already done, did not review your latest changes too detailed)

No, haven't made a lot of progress with this PR in the last weeks. Will do some more soon :).

boldtrn commented:
> Some minor changes I had to do or would like to have :) (maybe already done, did not review your latest changes too detailed)

These changes are now committed.

boldtrn commented:
My last commit is broken. I created a more generic approach for the spatial rule. Is this better? Should I change anything? 

BTW: the hack with the Object and the second method wrapping it into generic was needed in my case. Maybe we find a way around this.

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRule.java:
```diff
@@ -29,12 +29,17 @@
+
     SpatialRule addBorder(Polygon polygon);
 
     String getCountryIsoA3Name();
```
I think getCountryIsoA3Name is also too specific e.g. imagine the use case from phil where you have circles around a city to define different areas for the speeds. Maybe we keep everything more specific for now and switch later to a more generic version? The problem I see is that although getObject is already very generic, the parameters could be limited ... or should we remove transportationMode e.g. I can image that sometimes this makes no sense. Maybe the best bet would be `getObject(key, ConfigMap, _default)` and we provide a fast 'converter' from ReaderWay to ConfigMap ... hmmh ... not clear. Similar problem like we have for #472

boldtrn replied:
Ok, for now I renamed the method to getUniqueName, so we are already more generic at this point. As you said, the rest can be changed once we have a better feeling for the use cases.

karussell commented on web/src/main/java/com/graphhopper/http/SpatialRuleLookupBuilder.java:
```diff
@@ -54,30 +55,80 @@ public static SpatialRuleLookup build() {
                 }
             }
+
+            Double minLon = null;
```
Here we could use the same approach like in PathWrapper.calcRouteBBox via
bbox = BBox.createInverse and then bbox.update(lat,lon)?

But not sure if it fits perfectly

boldtrn replied:
That's a good idea. I updated the intersection code. Looks way better now.

karussell commented:
One more thing: I'm not sure if we should put a dependency inside DataFlagEncoder. Is there an alternative to feed it from outside somehow via the post processing hook, looping over all edges and setSpeed?

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRuleLookup.java:
```diff
@@ -0,0 +1,60 @@
+ *
+ * @author Robin Boldt
+ */
+public interface SpatialRuleLookup {
```
Here every method also needs some javadocs. And is visualize important?

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/countries/AustriaSpatialRule.java:
```diff
@@ -0,0 +1,56 @@
+ *
+ * @author Robin Boldt
+ */
+public class AustriaSpatialRule extends DefaultSpatialRule {
```
Are here some `@Override` missing?

boldtrn replied:
You are right, added them.

boldtrn commented:
Thanks for your feedback. I will change the code accordingly.

> Thanks & looks good. Just the question regarding specific vs. generic. Probably we should keep the specific stuff until we have a better impression of more use cases?

Yes, I have the same feeling right now. It seems a bit like premature optimization. Let's start easy and make it more generic when we need it :).

boldtrn commented:
I changed the the way the Spatial Lookup works a bit. 

When importing, an id is written for every edge. The id references to a spatial rule. In the GenericWeighting  these rules are evaluated.

Currently, the evaluation in the GenericWeighting is not yet perfect.

boldtrn commented on core/src/test/java/com/graphhopper/routing/weighting/GenericWeightingTest.java:
```diff
@@ -108,6 +112,43 @@ public void testCalcTime() {
     }
 
     @Test
+    public void testSpatial() {
```
I don't really like this test. It is rather long for a simple test.

karussell commented:
I have some notes and will further think about how we should progress here.

E.g. the singleton SpatialRuleRegister won't work pretty and is an unnecessary restriction. In general singleton is considered dangerous and for GraphHopper this cannot be used as we allow (for purpose) multiple instances of everything. Also testing is ugly for this etc.

The EMPTY_RULE_CONTAINER is potential buggy as you can add (by accident) multiple rules (at least I do not see a restriction)

Furthermore there are many javadocs missing, see the discussion in #891. But again, first let us decide on what we really want&need :) !

boldtrn commented:
> E.g. the singleton SpatialRuleRegister won't work pretty and is an unnecessary restriction.

Ok sure, I can change that.

> The EMPTY_RULE_CONTAINER is potential buggy as you can add (by accident) multiple rules (at least I do not see a restriction)

Will have a look, thanks.

> Furthermore there are many javadocs missing

Yes, this is a bit on purpose. I will do that before we merge this PR. When we constantly change the interface, the documentation might suffer.

>  But again, first let us decide on what we really want&need :) !

Good plan!

boldtrn commented:
I removed the Singleton. I think this was a good idea. This allows us to easily inject the config to make the CountryRules configureable. The configuring part is not done yet.

boldtrn commented:
> The EMPTY_RULE_CONTAINER is potential buggy as you can add (by accident) multiple rules (at least I do not see a restriction)

Good find! I updated the EMPTY_RULE_CONTAINER and merged the latest master.

karussell commented:
It looks better but still there is a bigger issue: the spatial rule index and the spatial rule registry are too loosely coupled. I think the registry stuff should be moved into the index. The problem is that the number of rules define the number of required bits in the encoded value, and so we

 - [ ] have to set the spatial rule index earlier (before passing it to the EncodingManager) and
 - [ ] ensure that the configured rules are always the same number as the loaded once. Otherwise we e.g. add two rules (need two bits) for the import and on load a different config applies and the encoded value requires only 1 bit leading to all strange things. So we either have to return the count in the details or add it to the version or something different to enforce this.  (This is a general problem with the DataFlagEncoder that will also be a problem for the recently introduced configurable height&width&weight - will mention it there too)

Furthermore:

 - [ ] the default for the index should be no rules at all and requiring 0 bits and the code is then skipped the lookups entirely
 - [ ] 'Destroy the SpatialRuleLookup' -> Something seems to be wrong. Because either the 'destroy' should be naturally done from the GC or we should keep it. Also import vs. load with different number of rules will break the encoder. Let me think about this.
 - [ ] SpatialRuleRegister: `public int getIdForName(String name)` should be `public int getId(SpatialRule rule)` instead
 - [ ] SpatialRuleRegister: `Map<Integer, SpatialRule>` -> just use an ArrayList :) ?
 - [x] It should be possible with the config properties to specify a list of countries as JSON files. Let us avoid this for now and only enable this feature via Java API for now.
 - [ ] And what do the access values 0, 5, accessMap.get("no") mean? (DataFlagEncoder.getEdgeValueForAccess)
 - [ ] Optionally move the builder class containing just one method into the index too

boldtrn commented:
> I think the registry stuff should be moved into the index. The problem is that the number of rules define the number of required bits in the encoded value

True, my idea for this was to completely get rid of the SpatialIndex, once we loaded the OSM file.

>  So we either have to return the count in the details or add it to the version or something different to enforce this.

This is a good idea. So I anyone loads a different graph than configured we can throw an exception.

> the default for the index should be no rules at all and requiring 0 bits and the code is then skipped the lookups entirely

Yes that is true, that is something I wanted to do when making it configureable.

>  'Destroy the SpatialRuleLookup' -> Something seems to be wrong. Because either the 'destroy' should be naturally done from the GC or we should keep it.

Well we pass the object reference to the GenericWeighting, we have to unreference it after the import I think.

>  just use an ArrayList :) ?

Probably, yes :)

> And what do the access values 0, 5, accessMap.get("no") mean? (DataFlagEncoder.getEdgeValueForAccess)

This was introduced when making track unaccessible. So the GenericWheighting tried to estimate how accessible a certain street is. So an accessible street would be 0. 5 was I think eventually accessible.

karussell commented on core/src/main/java/com/graphhopper/routing/util/DataFlagEncoder.java:
```diff
@@ -405,8 +394,23 @@ public long handleWayTags(ReaderWay way, long allowed, long relationFlags) {
+                    // block access if rule says so?
+//                    if (rule.getAccessible(way.getTag("highway"), "", AccessValue.ACCESSIBLE).
+//                            equals(AccessValue.NOT_ACCESSIBLE)) {
+//                        flags = accessEncoder.setValue(flags, NOT_ACCESSIBLE_KEY);
```
Should we store that access is not possible due to the rule here?

karussell commented on core/src/main/java/com/graphhopper/routing/util/DataFlagEncoder.java:
```diff
@@ -793,17 +799,33 @@ public boolean isStoreWidth() {
     @Override
     public int getVersion() {
-        return 1;
+        return 2;
```
increase due to access bits changed (3 -> 2)

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRuleContainer.java:
```diff
@@ -17,51 +17,56 @@
-        for (SpatialRule rule: this.rules) {
-            container.addRule(rule);
-        }
+        container.addRules(this.rules);
```
Another probably more explicit way would be to use a constructor which takes another container object instead?

karussell commented on core/src/test/java/com/graphhopper/routing/util/DataFlagEncoderTest.java:
```diff
@@ -148,19 +148,6 @@ public void testHighwaySpeed() {
-
-        way.setTag("vehicle", "no");
-        assertEquals(AccessValue.NOT_ACCESSIBLE, encoder.getEdgeAccessValue(encoder.handleWayTags(way, encoder.acceptWay(way), 0)));
-    }
```
will revert this removal ...

karussell commented:
Have now pushed a change suggestion.

BTW: The same config is now ensured via the spatial_rules number. The height&width etc config are similarly ensured already.

boldtrn commented on core/src/main/java/com/graphhopper/routing/util/DataFlagEncoder.java:
```diff
@@ -793,17 +799,33 @@ public boolean isStoreWidth() {
+
+    public DataFlagEncoder setSpatialRuleLookup(SpatialRuleLookup spatialRuleLookup) {
+        if (spatialRuleLookup.size() != spatialRules)
+            throw new IllegalArgumentException("You have to configure the encoder to accept the identical amount of rules " +
```
Should we get the number of spatial rules from the lookup then? But the lookup isn't available when constructing the encoder?

karussell replied:
Exactly that can be the problem

boldtrn commented on core/src/main/java/com/graphhopper/routing/util/DataFlagEncoder.java:
```diff
@@ -793,17 +799,33 @@ public boolean isStoreWidth() {
                 "|store_weight=" + storeWeight +
-                "|store_width=" + storeWidth;
+                "|store_width=" + storeWidth +
+                "|spatial_rules=" + spatialRules;
```
Should we safe the order and name of the countries? Or would this be too much overhead?

karussell replied:
For now lets keep it simple and a bit unsafer until we have something better :)

boldtrn commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRule.java:
```diff
@@ -17,21 +17,19 @@
+    double getMaxSpeed(String highway, double _default);
 
-    AccessValue isAccessible(String highwayTag, String transportationMode, AccessValue _default);
+    AccessValue getAccessible(String highwayTag, String transportationMode, AccessValue _default);
```
getAccessValue or getAccessibility?

karussell replied:
Ah, yes getAccessValue

boldtrn commented on web/src/main/java/com/graphhopper/http/DefaultModule.java:
```diff
@@ -48,15 +59,33 @@ public GraphHopper getGraphHopper() {
+        GHJson ghJson = new GHJsonBuilder().create();
+        JsonFeatureCollection jsonFeatureCollection = ghJson.fromJson(reader, JsonFeatureCollection.class);
+        return new SpatialRuleLookupBuilder().build(
+                Arrays.asList(new GermanySpatialRule(), new AustriaSpatialRule()),
```
Should we make the Countries configureable?

karussell replied:
We could do it with reflection? Maybe later?

boldtrn commented on core/src/main/java/com/graphhopper/routing/weighting/GenericWeighting.java:
```diff
@@ -124,40 +119,16 @@ public double calcWeight(EdgeIteratorState edgeState, boolean reverse, int prevO
         if (time == Long.MAX_VALUE)
             return Double.POSITIVE_INFINITY;
 
-        int spatialId = gEncoder.getSpatialId(edgeState.getFlags());
```
Did I missed it or is the SpatialIndex actually unused now? We safe the CountryId but don't evaluate it.

karussell replied:
This is true, yes. But we could apply speed and access changes when feeding here: https://github.com/boldtrn/graphhopper/blob/59b7e435f11628507cdbc254d27f7ceb3b31501d/core/src/main/java/com/graphhopper/routing/util/DataFlagEncoder.java#L408

I would avoid putting the index in the GenericWeighting ... too many things already there and as we could change speed and accessibility when feeding this is also not necessary for now

karussell replied:
Would you mind to add applying the speed and accessibility changes and recover may some tests that I removed? This seems indeed important to me, now that you pointed me to it

boldtrn replied:
> Would you mind to add applying the speed and accessibility changes and recover may some tests that I removed? This seems indeed important to me, now that you pointed me to it

So should I put it back into the GenericWeighting? Or should we keep it in the DataFlagEncoder?

karussell replied:
Please for now without GenericWeighting. So like I linked: determine the country and apply a speed reduction if necessary (similar with the access).

boldtrn replied:
> Would you mind to add applying the speed and accessibility changes and recover may some tests that I removed? This seems indeed important to me, now that you pointed me to it

Added these back to the DataFlagEncoder.

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRuleLookupArray.java:
```diff
@@ -20,20 +20,18 @@
  * The covered area is indexed as tiles, with every tile being "quadratic" having the same degree length on every side.
- * <p>
- * Currently one one rule per Tile is allowed. The later added rule will be added to the tile.
+ * Currently only a limited amount of rule combinations per tile are allowed.
```
It is not limited to the number of rules per container but to the number of rule combinations. E.g. A,B is one combination then A,B,D is another and so on. So maybe not that clear description?

boldtrn replied:
But is it really limited per tile? It is limited for the whole lookup, but not per tile.

karussell replied:
Ah, yes, it is the whole lookup

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRuleLookupArray.java:
```diff
@@ -53,15 +53,34 @@
+    SpatialRuleLookupArray(BBox bounds, double resolution, boolean exact) {
+        if (bounds == null)
+            throw new IllegalArgumentException("BBox cannot be null");
+        if (resolution < 1e-100)
```
The problem is resolution<=0 is not what we want and maybe we can raise this limit here a bit but not lower it IMO

boldtrn replied:
I am perfectly fine with the limit, I was just wondering why it is that number :).

karussell replied:
I just picked a very small one - maybe we can increase it with some assumption

boldtrn replied:
No, that's fine. Was just wondering.

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRuleLookupArray.java:
```diff
@@ -74,22 +93,13 @@ private int getNumberOfXGrids() {
-            return EMPTY_RULE;
-        }
+        if (lon < bounds.minLon || lon > bounds.maxLon || lat < bounds.minLat || lat > bounds.maxLat)
+            return EMPTY;
```
As it is in SpatialRule you can always write SpatialRule.EMPTY ?

boldtrn replied:
Yes, SpatialRule.EMPTY looks good I think. It's just a bit hard to read.

karussell replied:
Please rename it to EMPTY_RULE if you like. I have no preference :)

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRuleLookupArray.java:
```diff
@@ -174,35 +181,72 @@ public void addRule(SpatialRule rule) {
-                            lookupArray[i][j] = (byte) ruleIndex;
+                            lookupArray[i][j] = (byte) ruleContainerIndex;
                         } else {
                             // Merge Rules
```
Extracted already a bit ... it is ok for now IMO

boldtrn replied:
Yes I saw that. Should be fine.

karussell commented:
(BTW: also please have a look at some comments we made at 'outdated' commits here in this PR :))

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRuleLookupArray.java:
```diff
@@ -105,10 +115,10 @@ public SpatialRule lookupRule(double lat, double lon) {
 
-    protected int getRuleIndex(int xIndex, int yIndex) {
+    protected int getRuleContainerIndex(int xIndex, int yIndex) {
         if (xIndex < 0 || xIndex > lookupArray.length) {
```
Well, usually ifs for boundary checks are very fast. For now I wouldn't care that much. But we could check the result (in lookupRule) for EMPTY_RULE_INDEX and return EMPTY instead of the call to ruleContainers etc?

boldtrn replied:
Yeah, let's just leave it as is. If we see any performance issues we can still tweak?

boldtrn commented on core/src/main/java/com/graphhopper/routing/util/DataFlagEncoder.java:
```diff
@@ -296,6 +325,10 @@ public long handleWayTags(ReaderWay way, long allowed, long relationFlags) {
+                maxSpeed = getSpatialRule(way).getMaxSpeed(way.getTag("highway", ""), maxSpeed);
+            }
             double fwdSpeed = parseSpeed(way.getTag("maxspeed:forward"));
             if (fwdSpeed < 0 || maxSpeed > 0 && maxSpeed < fwdSpeed)
```
Is fwdSpeed > maxSpeed an iillegal/unwanted state? From a mapping perspective it shouldn't matter, right?

karussell replied:
It should be allowed, yes.

boldtrn commented:
Ok, so I updated this PR. Are we good to merge it or is there anything we should add/change?

karussell commented on core/src/main/java/com/graphhopper/routing/util/DataFlagEncoder.java:
```diff
@@ -418,6 +427,25 @@ public long handleWayTags(ReaderWay way, long allowed, long relationFlags) {
+        return false;
+    }
+
+    private SpatialRule getSpatialRule(ReaderWay way) {
```
Where is this method used?

karussell commented on core/src/test/java/com/graphhopper/routing/util/DataFlagEncoderTest.java:
```diff
@@ -346,7 +345,11 @@ public void testSpatialId() {
+        assertEquals(index.getSpatialId(SpatialRule.EMPTY), encoder.getSpatialId(e2.getFlags()));
+
+        assertEquals(AccessValue.EVENTUALLY_ACCESSIBLE, encoder.getAccessValue(e1.getFlags()));
+        assertEquals(AccessValue.ACCESSIBLE, encoder.getAccessValue(e2.getFlags()));
```
Would you mind to add a further test where the max_speed value is overwritten by the one stored in the index?

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRule.java:
```diff
@@ -0,0 +1,76 @@
+    /**
+     * Get the unique name for this rule, e.g. the ISO name of the country.
+     */
+    String getUniqueName();
```
rename to getId?

boldtrn replied:
Yes, at first I did not want to call it id, but this should be easier to understand and also imply that it should be unique. But I think it makes sense to rename it.

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRuleLookupBuilder.java:
```diff
@@ -0,0 +1,81 @@
+            }
+        }
+
+        if (polygonBounds.minLon == Double.MAX_VALUE) {
```
here it is safer to use !polygonBounds.isValid() instead

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRuleLookupArray.java:
```diff
@@ -0,0 +1,273 @@
+            }
+            System.out.println();
+        }
+    }
```
As you said this is no longer used/needed: would you mind to remove it?

boldtrn commented on core/src/main/java/com/graphhopper/routing/util/DataFlagEncoder.java:
```diff
@@ -82,9 +86,13 @@
     private EncodedValue highwayEncoder;
     private EncodedValue transportModeEncoder;
     private EncodedValue accessEncoder;
+    private static int NOT_ACCESSIBLE_KEY;
```
That variable wasn't used :)

boldtrn replied:
Somehow I cannot comment on your comment below.

> Can it be removed?

Yes, but it is used two times in the code. This would most probably lead to duplicated code. So I wouldn't want to do this, only if we have to?

karussell replied:
> Yes, but it is used two times in the code. This would most probably lead to duplicated code. So I wouldn't want to do this, only if we have to?

Ok

boldtrn commented on core/src/main/java/com/graphhopper/routing/util/DataFlagEncoder.java:
```diff
@@ -296,15 +325,19 @@ public long handleWayTags(ReaderWay way, long allowed, long relationFlags) {
+                maxSpeed = getSpatialRule(way).getMaxSpeed(way.getTag("highway", ""), maxSpeed);
+            }
             double fwdSpeed = parseSpeed(way.getTag("maxspeed:forward"));
-            if (fwdSpeed < 0 || maxSpeed > 0 && maxSpeed < fwdSpeed)
```
Yes you are right. Added another test.

boldtrn commented:
I updated this PR and added JavaDoc to the Interfaces. I realized that passing a string as TransportationMode is not a good idea and changed it to be an enum (sorry for the last minute change). I feel like it makes more sense like that. This enum could be used in the DataFlagEncoder to differ between different TransportationModes as well.

karussell commented:
Nice - Thanks!

karussell commented on core/src/main/java/com/graphhopper/routing/util/spatialrules/Polygon.java:
```diff
@@ -0,0 +1,182 @@
+ * This class represents a polygon that is defined by a set of points.
+ * Every point i is connected to point i-1 and i+1.
+ * <p>
+ * TODO: Howto design inner rings in the polygon?
```
Or we do this here #956
